cmd/mailserver-canary/main.go

```diff
@@ -0,0 +1,387 @@
+					"to":` + strconv.FormatInt(to.Unix(), 10) + `,
+					"limit": ` + fmt.Sprintf("%d", limit) + `,
+					"cursor": "` + cursor + `"
+		}]
+	}`)
+	reqMessagesResp := baseRPCResponse{}
+	err := json.Unmarshal([]byte(resp), &reqMessagesResp)
+	logger.Info(resp)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	if reqMessagesResp.Error != nil {
+		return common.Hash{}, createErrorFromJSONError(reqMessagesResp.Error)
+	}
+
+	switch hash := reqMessagesResp.Result.(type) {
+	case string:
+		if !strings.HasPrefix(hash, "0x") {
+			return common.Hash{}, errors.New("missing 0x prefix in shhext_requestMessages RPC call response")
+		}
+		b, err := hex.DecodeString(hash[2:])
+		return common.BytesToHash(b), err
+	default:
+		err = fmt.Errorf("expected a hash, received %+v", reqMessagesResp.Result)
+	}
+
+	return common.Hash{}, err
+}
+
+func joinPublicChat(w *whisper.Whisper, rpcClient *rpc.Client, name string) (string, whisper.TopicType, string, error) {
+	keyID, err := w.AddSymKeyFromPassword(name)
+	if err != nil {
+		return "", whisper.TopicType{}, "", err
+	}
+
+	h := sha3.NewKeccak256()
+	_, err = h.Write([]byte(name))
+	if err != nil {
+		return "", whisper.TopicType{}, "", err
+	}
+	fullTopic := h.Sum(nil)
+	topic := whisper.BytesToTopic(fullTopic)
+
+	filterID, err := createGroupChatMessageFilter(rpcClient, keyID, topic.String())
+
+	return keyID, topic, filterID, err
+}
+
+// createGroupChatMessageFilter create message filter with symmetric encryption.
+func createGroupChatMessageFilter(rpcCli *rpc.Client, symkeyID string, topic string) (string, error) {
+	resp := rpcCli.CallRaw(`{
+			"jsonrpc": "2.0",
+			"method": "shh_newMessageFilter", "params": [
+				{"type": "sym", "symKeyID": "` + symkeyID + `", "topics": [ "` + topic + `"], "allowP2P":true}
+			],
+			"id": 1
+		}`)
+
+	msgFilterResp := returnedIDResponse{}
+	err := json.Unmarshal([]byte(resp), &msgFilterResp)
+	messageFilterID := msgFilterResp.Result
+	if err == nil && msgFilterResp.Error != nil {
+		err = createErrorFromJSONError(msgFilterResp.Error)
+	}
+	return messageFilterID, err
+}
+
+func createErrorFromJSONError(jsonError interface{}) error {
+	errorObjMap := jsonError.(map[string]interface{})
+	if errorObjMap != nil {
+		return fmt.Errorf("error %d: %v", errorObjMap["code"], errorObjMap["message"])
+	}
+	return nil
+}
+
+func waitForMailServerResponse(events chan whisper.EnvelopeEvent, requestID common.Hash, timeout time.Duration) (*whisper.MailServerResponse, error) {
+	timeoutTimer := time.NewTimer(timeout)
+	for {
+		select {
+		case event := <-events:
+			if event.Hash == requestID {
+				resp, err := decodeMailServerResponse(event)
+				if resp != nil || err != nil {
+					timeoutTimer.Stop()
+					return resp, err
+				}
+			}
+		case <-timeoutTimer.C:
+			return nil, errors.New("timed out waiting for mailserver response")
+		}
+	}
+}
+
+func decodeMailServerResponse(event whisper.EnvelopeEvent) (response *whisper.MailServerResponse, err error) {
```

Comment:
I wasn't being strict on unexpected events. We could return a `errors.New("unexpected event type")` in that case, but not sure that we gain anything by breaking the functionality by default in case a new event type is added in the future.